john.conf: add external "wordcase" mode

rofl0r opened the pull request:
expands an input phrase, e.g. "foo bar" into all possible combinations
with words separated by space capitalized or not:
foo bar
Foo bar
Foo Bar
foo Bar

the code is identical to 'case' external mode, with the addition of one
condition in the new() function.

rofl0r committed:
john.conf: add external "wordcase" mode

expands an input phrase, e.g. "foo bar" into all possible combinations
with words separated by space capitalized or not:

foo bar
Foo bar
Foo Bar
foo Bar

since the code is identical to `Case` external mode, with the addition of
one condition in the new() function, the case mode was split into an
External_base shared section included from both Case and Wordcase mode,
using the added variable `word_mode` to either always execute the block
in question in case of `Case` mode, or only when the conditions
interesting for `Wordcase` apply.

solardiz commented:
> the code is identical to 'case' external mode, with the addition of one
> condition in the new() function.

Maybe we should avoid code duplication between these modes by using `.include` of a shared section?

@magnumripper What do you think? I think you "pioneered"  that approach with some other modes.

magnumripper commented:
> @magnumripper What do you think? I think you "pioneered" that approach with some other modes.

It sometimes makes the code harder to read/follow, but as long as whole functions are re-used (like in this case I presume) it might be a good idea.

rofl0r commented:
i don't know about the abilities of john's "C script", but if it would work to call a named function
we could basically rename new in external:case to new_with_mode(int mode)
and then make eternal:case have new() that just calls new_with_mode(0) and wordcase new_with_mode(1) and then just enable/disable the additional branch depending on whether mode is set.

magnumripper commented:
> i don't know about the abilities of john's "C script", but if it would work to call a named function
> we could basically rename new in external:case to new_with_mode(int mode)

Unfortunately that's not supported; https://raw.githubusercontent.com/openwall/john/bleeding-jumbo/doc/EXTERNAL

What you could do is what Jim (I think) did with `List.External_base:Double` and variant functions such as `[List.External:Double]`: You could just have different `init()` for the original "case" and your new mode (introducing a new variable instead of using a function argument).

rofl0r commented:
> What you could do is what Jim (I think) did with List.External_base:Double and variant functions such as [List.External:Double]: You could just have different init() for the original "case" and your new mode (introducing a new variable instead of using a function argument).

good advice, thanks. updated PR accordingly.

rofl0r commented on run/john.conf:
```diff
@@ -4008,18 +4003,24 @@ void new()
+		   mode added we execute it either always when word_mode isn't
+		   used or at the beginning of a word or when the previous char
+		   was space */
+		if (word_mode == 0 || wlen == 0 || word[wlen-1] == 0x20) {
```
this is the new condition, however it relies on short-circuit evaluation assuming that when wlen == 0 the test word[wlen-1] == 0x20 is skipped; so this might introduce an out-of-bounds memory access. any ideas how to best fix this ?

solardiz replied:
Conservative changes just to this place:

```c
int prevch;
prevch = ' ';
if (wlen) prevch = word[wlen - 1];
if (!word_mode || prevch == ' ') {
```

A better fix would probably be to initialize prevch out of the loop and to add `prevch = ch;` near the `++wlen;` line.

BTW, `!word_mode` is a bit quicker than `word_mode == 0`. (The external mode compiler is mostly non-optimizing.)

rofl0r replied:
thanks! pushed a new commit addressing this, i made it into a separate commit because i thought this might be interesting for other log readers, but feel free to squash as desired.

rofl0r committed:
john.conf: fix out-of-bounds access in external:case/wordcase

because john's interpreter doesn't short-circuit evalutations, we
need to rewrite the condition

    word_mode == 0 || wlen == 0 || word[wlen-1] == 0x20

by using an additional variable in order not to access word[-1].

solardiz reviewed (approved):
Looks OK'ish to me, although I didn't test it.

rofl0r commented:
thanks, i did some quick testing and it seemed correct. while i have you around would you mind telling me where the code for the external compiler/vm is located? it seems to perform pretty well so i'd like to study its design.

solardiz commented:
> where the code for the external compiler/vm is located?

It's in `compiler.c`.

While it's mostly non-optimizing, there are a couple of optional optimizations: the use of GNU C "Labels as Values" (when building with GCC or compatible) and the existence/generation of multi-push VM instructions (a way to lower VM overhead by doing more work per instruction). So the basic logic implementing the same functionality (but slower) doesn't require all of the code that's present in the source file. Yet even with those optimizations, the file is pretty small (for what it does) and I hope will be fun for you to study. And yes, I wrote it, so I'm flattered by your interest in it.
